runtime/src/dpe_crypto.rs

```diff
@@ -78,64 +101,171 @@ impl<'a> Crypto for DpeCrypto<'a> {
 
     fn derive_cdi(
         &mut self,
-        _algs: AlgLen,
-        _measurement: &Digest,
-        _info: &[u8],
+        algs: AlgLen,
+        measurement: &Digest,
+        info: &[u8],
     ) -> Result<Self::Cdi, CryptoError> {
-        Err(CryptoError::NotImplemented)
-    }
+        match algs {
+            AlgLen::Bit256 => Err(CryptoError::Size),
+            AlgLen::Bit384 => {
+                let mut hasher = self
+                    .hash_initialize(algs)
+                    .map_err(|_| CryptoError::HashError)?;
+                hasher
+                    .update(measurement.bytes())
+                    .map_err(|_| CryptoError::HashError)?;
+                hasher.update(info).map_err(|_| CryptoError::HashError)?;
+                let context = hasher.finish().map_err(|_| CryptoError::HashError)?;
 
-    fn derive_private_key(
-        &mut self,
-        _algs: AlgLen,
-        _cdi: &Self::Cdi,
-        _label: &[u8],
-        _info: &[u8],
-    ) -> Result<Self::PrivKey, CryptoError> {
-        Err(CryptoError::NotImplemented)
+                hmac384_kdf(
+                    self.hmac384,
+                    KeyReadArgs::new(KEY_ID_RT_CDI).into(),
+                    b"derive_cdi",
+                    Some(context.bytes()),
+                    self.trng,
+                    KeyWriteArgs::new(KEY_ID_DPE_CDI, KeyUsage::default().set_hmac_data_en())
+                        .into(),
+                )
+                .map_err(|_| CryptoError::CryptoLibError)?;
+                Ok(KEY_ID_RT_CDI)
+            }
+        }
     }
 
-    fn derive_ecdsa_pub(
+    fn derive_key_pair(
         &mut self,
-        _algs: AlgLen,
-        _priv_key: &Self::PrivKey,
-    ) -> Result<EcdsaPub, CryptoError> {
-        Err(CryptoError::NotImplemented)
+        algs: AlgLen,
+        cdi: &Self::Cdi,
+        label: &[u8],
+        info: &[u8],
+    ) -> Result<(Self::PrivKey, EcdsaPub), CryptoError> {
+        match algs {
+            AlgLen::Bit256 => Err(CryptoError::Size),
+            AlgLen::Bit384 => {
+                hmac384_kdf(
+                    self.hmac384,
+                    KeyReadArgs::new(*cdi).into(),
+                    label,
+                    Some(info),
+                    self.trng,
+                    KeyWriteArgs::new(
+                        KEY_ID_DPE_CDI,
+                        KeyUsage::default().set_ecc_key_gen_seed_en(),
+                    )
```

Comment:
This has a bit of a side-effect in that the input CDI is being modified. Maybe we should use another KV slot as scratch space (we can clear it before exiting the function)